frex/utils/constraint_solver.py

```diff
@@ -0,0 +1,208 @@
+from __future__ import annotations
+from ortools.linear_solver import pywraplp
+from frex.models import Candidate, ConstraintSectionSolution, ConstraintSolution
+from typing import Tuple
+from frex.utils.common import rgetattr
+from frex.utils import ConstraintType, Constraint
+from enum import Enum
+
+
+class ConstraintSolver:
+    """
+    A class to perform constraint solving to produce a final solution of items using constraints on the overall
+    set of items.
+    """
+
+    def __init__(self):
+        self.solver = pywraplp.Solver.CreateSolver("SCIP")
+        self.candidates = ()
+        self.sections = 1
+        self.per_section_count = 1
+        self.section_constraints = []
+        self.overall_constraints = []
+
+    def set_candidates(self, *, candidates: Tuple[Candidate]):
+        """
+        Set the candidates that will be used to produce the solution.
+        Candidates are expected to be produced as the output of some pipeline, which handles scoring.
+        The solver will not handle any sort of scoring for the candidates, but rather it will produce an optimized
+        solution based on the Candidates' total_score (which should be computed by a pipeline) and other constraints.
+
+        :param candidates: A tuple of Candidate objects, with corresponding domain_objects and scores
+        :return: self, with an updated list of candidates
+        """
+        self.candidates = candidates
+        return self
+
+    def set_sections(self, *, num_sections: int):
+        """
+        Set the number of sections in the solution. A section can be thought of as a grouping of items, and the solution
```

Comment:
Either call the variable sections or make the setter set_num_sections, either way be consistent.